Source/Diagnostics/ReducedDiags/ParticleHistogram.cpp

```diff
@@ -0,0 +1,167 @@
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "ParticleHistogram.H"
+#include "WarpX.H"
+#include "Utils/WarpXConst.H"
+#include "Utils/WarpXUtil.H"
+#include <AMReX_REAL.H>
+#include <AMReX_ParticleReduce.H>
+#include <limits>
+
+using namespace amrex;
+
+// constructor
+ParticleHistogram::ParticleHistogram (std::string rd_name)
+: ReducedDiags{rd_name}
+{
+
+    // read parameters
+    ParmParse pp(rd_name);
+    std::string selected_species_name;
+    pp.get("species",selected_species_name);
+    pp.get("bin_number",m_bin_num);
+    pp.get("bin_max",   m_bin_max);
+    pp.get("bin_min",   m_bin_min);
+    pp.query("normalization",m_norm);
+    m_bin_size = (m_bin_max - m_bin_min) / m_bin_num;
+    std::string function_string = "";
+    Store_parserString(pp,"histogram_function(t,x,y,z,ux,uy,uz)",
+                       function_string);
+    m_parser.reset(new ParserWrapper<7>(
+        makeParser(function_string,{"t","x","y","z","ux","uy","uz"})));
+
+    // get MultiParticleContainer class object
+    auto & mypc = WarpX::GetInstance().GetPartContainer();
+    // get species names (std::vector<std::string>)
+    auto const species_names = mypc.GetSpeciesNames();
+    // select species
+    for ( int i = 0; i < mypc.nSpecies(); ++i )
+    {
+        if ( selected_species_name == species_names[i] )
+        { m_selected_species_id = i; }
+    }
+    // if m_selected_species_id is not modified
+    if ( m_selected_species_id == -1 )
+    { Abort("Unknown species for ParticleHistogram reduced diagnostic."); }
+
+    // resize data array
+    m_data.resize(m_bin_num,0.0);
+
+    if (ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs;
+            ofs.open(m_path + m_rd_name + "." + m_extension,
+                std::ofstream::out | std::ofstream::app);
+            // write header row
+            ofs << "#";
+            ofs << "[1]step()";
+            ofs << m_sep;
+            ofs << "[2]time(s)";
+            for (int i = 0; i < m_bin_num; ++i)
+            {
+                ofs << m_sep;
+                ofs << "[" + std::to_string(3+i) + "]";
+                Real b = m_bin_min + m_bin_size*(Real(i)+0.5);
+                ofs << "bin" + std::to_string(1+i)
+                             + "=" + std::to_string(b) + "()";
+            }
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+
+}
+// end constructor
+
+// function that computes the histogram
+void ParticleHistogram::ComputeDiags (int step)
+{
+
+    // Judge if the diags should be done
+    if ( (step+1) % m_freq != 0 ) return;
+
+    // get time at level 0
+    auto const t = WarpX::GetInstance().gett_new(0);
+
+    // get MultiParticleContainer class object
+    auto & mypc = WarpX::GetInstance().GetPartContainer();
+
+    // get WarpXParticleContainer class object
+    auto & myspc = mypc.GetParticleContainer(m_selected_species_id);
```

Comment:
```suggestion
    // get MultiParticleContainer class object
    auto const & mypc = WarpX::GetInstance().GetPartContainer();

    // get WarpXParticleContainer class object
    auto const & myspc = mypc.GetParticleContainer(m_selected_species_id);
```
Could these work? I think you mentioned in another PR That one of them doesn't, but I'm not sure.

Reply:
I tried and seem like the second one can work, the first one can not.
I will do the second one.